jupyter-spark/extensions/spark.js

```diff
@@ -0,0 +1,136 @@
+var api = "/spark/api/v1";
+var update_frequency = 10000; // ms
+
+
+/* 
+cache is an array of application objects with an added property jobs.
+application.jobs is the result of the /applications/applicationId/jobs
+API request.
+*/
+var cache = [];
+
+var update = function() {
+    update_cache(update_dialog_contents);
+}
+
+// callbacks follows jQuery callback style, can be either single function or array of functions
+// callbacks will be passed the cache as a parameter
+var update_cache = function(callbacks) {
+    var cbs;
+    if (callbacks) {
+        cbs = $.Callbacks();
+        cbs.add(callbacks);
+    }
+    $.getJSON(api + '/applications').done(function(applications) {
+        applications.forEach(function(application, i) {
+            $.getJSON(api + '/applications/' + application.id + '/jobs').done(function (jobs) {
+                cache[i] = application;
+                cache[i].jobs = jobs;
+                if (cbs) {
+                    // FIXME: need to check if all applications have been updated
+                    cbs.fire(cache);
+                }
+            });
+        });
+    });
+}
+
+var update_dialog_contents = function() {
+    if ($('#dialog_contents').length) {
+        var element = $('<div/>').attr('id', 'dialog_contents');
+        cache.forEach(function(application, i){
+            element.append(create_application_table(application));
+        })
+
+        $('#dialog_contents').replaceWith(element);
+    }
+}
+
+var create_application_table = function(e) {
+    var application_div = $('<div/>');
+    application_div.append($('<h5/>').text(e.name + ': ' + e.id));
+    var application_table = $('<table/>').addClass('table table-hover');
+
+    var header_row = $('<tr/>');
+    header_row.append($('<th/>').text('Job ID'));
+    header_row.append($('<th/>').text('Job Name'));
+    header_row.append($('<th/>').text('Progress'));
+    application_table.append(header_row);
+
+    e.jobs.forEach(function(job, i) {
+        application_table.append(create_table_row(job));
+    });
+
+    application_div.append(application_table);
+    return application_div;
+}
+
+var create_table_row = function(e) {
+    var row = $('<tr/>');
+    row.append($('<td/>').text(e.jobId));
+    row.append($('<td/>').text(e.name));
+    
+    var status_class;
+    switch(e.status) {
+        case 'SUCCEEDED':
+            status_class = 'progress-bar-success';
+            break;
+        case 'RUNNING':
+            status_class = 'progress-bar-info';
+            break;
+        case 'FAILED':
+            status_class = 'progress-bar-danger';
+            break;
+        case 'UNKNOWN':
+            status_class = 'progress-bar-warning';
+            break;
+    }
+    
+    // progress defined in percent
+    var progress = e.numCompletedTasks / e.numTasks * 100;
```

Comment:
Is it possible to get a divide-by-zero here? If so, we should guard against it. If not, all is well!